### PR TITLE
[GP MODEL]  integrating kafka

### PR DESCRIPTION
**Pull request by NoahNsimbe**
**_WHAT DOES THIS PR DO?_**

- [x] Links GP model to message broker

**_WHAT ISSUES ARE RELATED TO THIS PR?_**

- Jira cards
    - []()
- GitHub issues
    - Closes #<enter GitHub issue number here>

**_HOW DO I TEST OUT THIS PR?_**

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 

**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**

**_ARE THERE ANY RELATED PRs?_**

### CODE REVIEW

File: .github/workflows/deploy-apis-to-production.yml
Comment (NoahNsimbe) [inline .github/workflows/deploy-apis-to-production.yml:1]: These changes need to be reviewed @123MwanjeMike

File: .github/workflows/deploy-apis-to-staging.yml
Comment (NoahNsimbe) [inline .github/workflows/deploy-apis-to-staging.yml:1]: These changes need to be reviewed @123MwanjeMike

File: k8s/gp-model/templates/deployment.yaml
Comment (NoahNsimbe) [inline k8s/gp-model/templates/deployment.yaml:1]: These changes need to be reviewed @123MwanjeMike

File: k8s/gp-model/templates/cronjob.yaml
Comment (NoahNsimbe) [inline k8s/gp-model/templates/cronjob.yaml:1]: These changes need to be reviewed @123MwanjeMike

File: src/gp-model/config.py [hunk 1]
```diff
@@ -1,19 +1,27 @@
...
     TESTING = False
     CSRF_ENABLED = True
     SECRET_KEY = os.getenv("SECRET_KEY")
     CREDENTIALS = os.getenv('GOOGLE_APPLICATION_CREDENTIALS')
-
+    MEASUREMENTS_TOPIC = os.getenv('MEASUREMENTS_TOPIC')
+    AIRQLOUDS_TOPIC = os.getenv('AIRQLOUDS_TOPIC')
+    BOOTSTRAP_SERVERS = os.getenv('BOOTSTRAP_SERVERS')
+    GP_MODEL_DB = os.getenv('GP_MODEL_DB')
+    GP_MODEL_DB_URI = os.getenv('GP_MODEL_DB_URI')
```
Comment (NoahNsimbe) [inline src/gp-model/config.py:24]: Secret manager needs to be updated @uman95

**Review by 123MwanjeMike — APPROVED** [review]
Great work @NoahNsimbe 
LGTM

File: src/gp-model/config.py (hunk 1 above)
Comment (123MwanjeMike) [reply src/gp-model/config.py:24]: Could you update those of prod @NoahNsimbe 
@uman95 may only have access to staging

File: src/gp-model/Dockerfile [hunk 2]
```diff
@@ -12,17 +12,17 @@ ENV PYTHONUNBUFFERED 1
 # production 
 FROM base as production 
 ENV ENV=production
-# run application
 CMD ["python", "main.py", "--tenant=airqo"]
 
-# production 
+# staing
```
Comment (Baalmart) [inline src/gp-model/Dockerfile:17]: *#staging

File: src/gp-model/config.py (hunk 1 above)
Comment (Baalmart) [reply src/gp-model/config.py:24]: Curious about the information required in AirQlouds.
When an AirQloud is just created, it does not have any Sites in it until is "refreshed".

I am curious as to how this subscription to AirQlouds topic will be utilised. @uman95 / @NoahNsimbe 

Thanks!

**Review by Baalmart — APPROVED** [review]
Thanks @NoahNsimbe , this is good work. I have also just left a few comments for your review. cc: @uman95

File: src/gp-model/message-broker.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                    try:
+                        print(f"\n\nReceived airqlouds: {data}\n\n")
+                        formatted_data = [
+                            {
+                                "_id": row["_id"],
+                                "long_name": row.get("long_name", ""),
+                                "network": row.get("network"),
+                                "admin_level": row.get("admin_level"),
+                                "name": row.get("name", ""),
+                                "sites": row.get("sites"),
```
Comment (uman95) [inline src/gp-model/message-broker.py:68]: Want to confirm again that "sites" here mean _site-id_ and not any other site identifier cos _site-id_ is what is needed.

File: src/gp-model/message-broker.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            try:
+                data = msg.value
+                topic = msg.topic
+
+                if topic == Config.MEASUREMENTS_TOPIC:
+                    print(f"\n\nReceived measurements: {data}\n\n")
+                    try:
+                        formatted_data = [
+                            {
+                                "site_id": row.get("site_id", None),
```
Comment (uman95) [inline src/gp-model/message-broker.py:35]: Nice, _site-id_ is what is actually needed.

File: src/gp-model/config.py (hunk 1 above)
Comment (uman95) [reply src/gp-model/config.py:24]: Thanks @Baalmart, the coordinates and sites under the AirQlouds are the most important fields required. As to newly created AirQlouds not having sites, we don't immediately build a spat-temp model (and consequently heatmap) on AirQlouds not even on newly added AirQlouds. It's more or less a request before building such model, which I believe such AirQloud must have been functional with available sites.

File: src/gp-model/message-broker.py (hunk 3 above)
Comment (NoahNsimbe) [reply src/gp-model/message-broker.py:68]: Okay @uman95 shall confirm that

File: src/gp-model/Dockerfile (hunk 2 above)
Comment (NoahNsimbe) [reply src/gp-model/Dockerfile:17]: Thanks @Baalmart let me fix this

File: src/gp-model/config.py (hunk 1 above)
Comment (NoahNsimbe) [reply src/gp-model/config.py:24]: Thanks @Baalmart for highlighting this. Let me review the implementation.

**Review by Baalmart — COMMENTED** [review]
@NoahNsimbe , merge conflicts have emerged

**Review by Baalmart — COMMENTED** [review]
@NoahNsimbe , merge conflicts have emerged

**Pull request merged**